File: app/src/main/java/io/zenandroid/greenfield/service/FlickrServiceImpl.java
package io.zenandroid.greenfield.service;

import android.support.annotation.Nullable;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import io.zenandroid.greenfield.api.FlickrApi;
import io.zenandroid.greenfield.model.ImageListResponse;
import io.zenandroid.greenfield.util.EspressoIdlingResource;

public class FlickrServiceImpl implements FlickrService {

	private final static String TAG = FlickrServiceImpl.class.getSimpleName();

	private final FlickrApi flickrApi;

	@Inject
	public FlickrServiceImpl(FlickrApi api) {
		flickrApi = api;
	}

	@Override
	public Single<ImageListResponse> getImageList(@Nullable String tags) {
		EspressoIdlingResource.getInstance().increment();
		return flickrApi.getPhotos(tags)
				.subscribeOn(Schedulers.computation())
				.observeOn(AndroidSchedulers.mainThread())
				.doFinally(() -> EspressoIdlingResource.getInstance().decrement());
	}
}

File: app/src/main/java/io/zenandroid/greenfield/service/FlickrServiceImpl.kt
package io.zenandroid.greenfield.service

import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.zenandroid.greenfield.api.FlickrApi
import io.zenandroid.greenfield.model.ImageListResponse
import io.zenandroid.greenfield.util.EspressoIdlingResource
import javax.inject.Inject

class FlickrServiceImpl @Inject
constructor(private val flickrApi: FlickrApi) : FlickrService {

    override fun getImageList(tags: String?): Single<ImageListResponse> {
        EspressoIdlingResource.increment()
        return flickrApi.getPhotos(tags)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally { EspressoIdlingResource.decrement() }
    }

    companion object {

        private val TAG = FlickrServiceImpl::class.java.simpleName
    }
}
